srv/src/main/java/my/bookshop/handlers/AdminServiceAddressHandler.java

```diff
@@ -0,0 +1,189 @@
+package my.bookshop.handlers;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.annotation.Resource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.sap.cds.Result;
+import com.sap.cds.Struct;
+import com.sap.cds.ql.CQL;
+import com.sap.cds.ql.Select;
+import com.sap.cds.ql.Upsert;
+import com.sap.cds.ql.cqn.CqnPredicate;
+import com.sap.cds.ql.cqn.CqnSelectListItem;
+import com.sap.cds.reflect.CdsModel;
+import com.sap.cds.services.ErrorStatuses;
+import com.sap.cds.services.ServiceException;
+import com.sap.cds.services.cds.CdsReadEventContext;
+import com.sap.cds.services.cds.CdsService;
+import com.sap.cds.services.draft.DraftPatchEventContext;
+import com.sap.cds.services.handler.EventHandler;
+import com.sap.cds.services.handler.annotations.Before;
+import com.sap.cds.services.handler.annotations.On;
+import com.sap.cds.services.handler.annotations.ServiceName;
+import com.sap.cds.services.messaging.TopicMessageEventContext;
+import com.sap.cds.services.persistence.PersistenceService;
+import com.sap.cloud.sdk.cloudplatform.resilience.ResilienceConfiguration;
+import com.sap.cloud.sdk.cloudplatform.resilience.ResilienceConfiguration.TimeLimiterConfiguration;
+import com.sap.cloud.sdk.cloudplatform.resilience.ResilienceDecorator;
+
+import cds.gen.adminservice.Addresses;
+import cds.gen.adminservice.AdminService_;
+import cds.gen.adminservice.Orders;
+import cds.gen.api_business_partner.ABusinessPartnerAddress;
+import cds.gen.api_business_partner.ABusinessPartnerAddress_;
+import cds.gen.api_business_partner.ApiBusinessPartner_;
+import cds.gen.api_business_partner.BOBusinessPartnerChanged;
+import cds.gen.my.bookshop.Addresses_;
+import my.bookshop.MessageKeys;
+
+/**
+ * Custom handler for the Admin Service Addresses, which come from a remote S/4 System
+ */
+@Component
+@ServiceName(AdminService_.CDS_NAME)
+public class AdminServiceAddressHandler implements EventHandler {
+
+	private final static Logger logger = LoggerFactory.getLogger(AdminServiceAddressHandler.class);
+
+	// We are mashing up the AdminService with two other services...
+	@Autowired
+	private PersistenceService db;
+
+	@Resource(name = ApiBusinessPartner_.CDS_NAME)
+	private CdsService bupa;
```

Comment:
Should we inject here through the constructor as Robin already suggested for our other handlers?

Reply:
yes, I can change it.